1_fetch/src/get_nwis_data.R

```diff
@@ -1,48 +1,91 @@
+  
+  # function to simultaneously download site information and data for set parameter
+  
+    # create the file name needed for download_nwis_site_data
+      download_file <- file.path(out_dir, paste0("nwis_", site_num, "_nwisdata.csv"))
+        
+    # download parameter data
+      val <- retry(download_nwis_site_data(download_file, startDate, endDate, parameterCd = parameterCd), maxErrors = 10, sleep = 2)
+        
+    # download site info
+      val <- retry(download_nwis_site_info(site_num, out_dir), maxErrors = 10, sleep = 2)
+        
+      return(out_dir)
+  
 }
 
-nwis_site_info <- function(fileout, site_data){
-  site_no <- unique(site_data$site_no)
-  site_info <- dataRetrieval::readNWISsite(site_no)
-  write_csv(site_info, fileout)
-  return(fileout)
-}
 
 
-download_nwis_site_data <- function(filepath, parameterCd = '00010', startDate="2014-05-01", endDate="2015-05-01"){
+download_nwis_site_info <- function(site_num, out_dir){ 
   
-  # filepaths look something like directory/nwis_01432160_data.csv,
-  # remove the directory with basename() and extract the 01432160 with the regular expression match
-  site_num <- basename(filepath) %>% 
-    stringr::str_extract(pattern = "(?:[0-9]+)")
+  # function to access and download site information for set site number
+      
+      site_info <- dataRetrieval::readNWISsite(site_num)
+      readr::write_csv(site_info, paste0(out_dir, "nwis_", site_num, "_siteinfo.csv"))
+      return(paste0(out_dir, "nwis_", site_num, "_siteinfo.csv"))
   
-  # readNWISdata is from the dataRetrieval package
-  data_out <- readNWISdata(sites=site_num, service="iv", 
-                           parameterCd = parameterCd, startDate = startDate, endDate = endDate)
+}
+
+
 
-  # -- simulating a failure-prone web-sevice here, do not edit --
-  set.seed(Sys.time())
-  if (sample(c(T,F,F,F), 1)){
-    stop(site_num, ' has failed due to connection timeout. Try tar_make() again')
-  }
-  # -- end of do-not-edit block
+download_nwis_site_data <- function(filepath, parameterCd, startDate, endDate){ 
   
-  write_csv(data_out, file = filepath)
-  return(filepath)
+  # function to access and download data for set parameter
+  
+      site_num <- basename(filepath) %>% 
+        stringr::str_extract(pattern = "(?:[0-9]+)")
+      
+      data_out <- dataRetrieval::readNWISdata(
+        sites=site_num, 
+        service="iv", 
+        parameterCd = parameterCd, 
+        startDate = startDate, 
+        endDate = endDate
+        )
+  
+    # -- simulating a failure-prone web-service here, do not edit --
+      set.seed(Sys.time())
+      if (sample(c(T,F,F,F), 1)){
+        stop(site_num, ' has failed due to connection timeout. Try tar_make() again')
+      }
+    # -- end of do-not-edit block
+    
+      readr::write_csv(data_out, filepath)
+      return(filepath)
+    
 }
 
+
+
+retry <- function(expr, isError=function(x) "try-error" %in% class(x), maxErrors, sleep) { 
```

Comment:
Oh wow! You implemented your own `retry()`!! Ahead of the curve for sure - this is something that we talk about handling later in this pipelines training, but nice job on trying to work on it now. You wrote a custom function, but there is actually a package that does this for us! We will introduce [a package called `retry`](https://cran.r-project.org/web/packages/retry/retry.pdf) in a later module.

Reply:
To be fair, I found that function on stackoverflow to implement. :) I should have figured there was a package!